examples/rcn.py

```diff
@@ -0,0 +1,392 @@
+ax.imshow(img[pad : 200 - pad, pad : 200 - pad], cmap="gray")
+
+
+# %% [markdown]
+# ## 3.1 Visualize the filters
+
+# %% [markdown]
+# The filters are used to detect the oriented edges on a given image. They are pre-computed using Gabor filters.
+
+# %%
+fig, ax = plt.subplots(4, 4, figsize=(10, 10))
+for i in range(filters.shape[0]):
+    idx = np.unravel_index(i, (4, 4))
+    ax[idx].imshow(filters[i], cmap="gray")
+    ax[idx].axis("off")
+
+fig.tight_layout()
+
+# %% [markdown]
+# # 4. Make pgmax graph
+
+# %% [markdown]
+# ## 4.1 Make variables
+
+# %%
+start = time.time()
+assert frcs.shape[0] == edges.shape[0]
+
+variables_all_models = {}
+for idx in range(frcs.shape[0]):
+    frc = frcs[idx]
+    variables_all_models[idx] = groups.NDVariableArray(
+        variable_size=M, shape=(frc.shape[0],)
+    )
+
+end = time.time()
+print(f"Creating variables took {end-start:.3f} seconds.")
+
+
+# %% [markdown]
+# ## 4.2 Make factors
+
+# %% [markdown]
+# ### 4.2.1 Pre-compute the valid configs for different perturb radii.
+
+# %%
+def valid_configs(r: int) -> np.ndarray:
+    """Returns the valid configurations for the potential matrix given the perturb radius.
+
+    Args:
+        r: Peturb radius
+
+    Returns:
+        A configuration matrix (shape n X 2) where n is the number of valid configurations.
+    """
+
+    configs = []
+    for i, (r1, c1) in enumerate(np.array(np.unravel_index(np.arange(M), (25, 25))).T):
+        r2_min = max(r1 - r, 0)
+        r2_max = min(r1 + r, 2 * hps)
+        c2_min = max(c1 - r, 0)
+        c2_max = min(c1 + r, 2 * vps)
+        j = np.ravel_multi_index(
+            tuple(np.mgrid[r2_min : r2_max + 1, c2_min : c2_max + 1]), (25, 25)
+        ).ravel()
+        configs.append(np.stack([np.full(j.shape, fill_value=i), j], axis=1))
+
+    return np.concatenate(configs)
+
+
+max_perturb_radius = 25
+valid_configs_list = [valid_configs(r) for r in range(max_perturb_radius)]
+
+# %% [markdown]
+# ### 4.2.2 Make the factor graph
+
+# %%
+start = end
+fg = graph.FactorGraph(variables=variables_all_models)
+for idx in range(edges.shape[0]):
+    edge = edges[idx]
+
+    for e in edge:
+        i1, i2, r = e
+        fg.add_factor(
+            [(idx, i1), (idx, i2)],
+            valid_configs_list[r],
+        )
+
+end = time.time()
+print(f"Creating factors took {end-start:.3f} seconds.")
+
+
+# %% [markdown]
+# # 5. Run inference
+
+# %% [markdown]
+# ## 5.1 Helper functions to initialize the evidence for a given image
+
+# %%
+def get_bu_msg(img: np.ndarray) -> np.ndarray:
+    """Computes the bottom-up messages given a test image.
+
+    Args:
+        img: The rgb image to compute bottom up messages on (H x W x 3).
+
+    Returns:
+        An array of shape [16 x H x W] denoting the presence or absence of an oriented line segment at a particular location.
+            The elements of this array belong to the set {+1, -1}.
+    """
+
+    num_orients = 16
+    brightness_diff_threshold = 40.0
```

Comment:
Again, these can be grouped and defined along with model params if neededm, but these are not arbitrary as they should match the model you loaded, so read it from the loaded model (filters). BDT can be but not num_orients

Reply:
Added configs at one place as suggested above.